src/v/io/cache.h

```diff
@@ -0,0 +1,662 @@
+ * As of the Change Date specified in that file, in accordance with
+ * the Business Source License, use of this software will be governed
+ * by the Apache License, Version 2.0
+ */
+#pragma once
+
+#include <boost/intrusive/list.hpp>
+#include <boost/intrusive/parent_from_member.hpp>
+
+#include <algorithm>
+#include <optional>
+
+namespace testing_details {
+class cache_hook_accessor;
+}; // namespace testing_details
+
+/**
+ * \defgroup cache Cache
+ *
+ * An implementation of the s3-fifo cache eviction algorithm.
+ *
+ * Yang, et al., "FIFO queues are all you need for cache eviction", SOSP '23,
+ * https://dl.acm.org/doi/10.1145/3600006.3613147
+ *
+ * The s3-fifo eviction algorithm uses three queues named small, main, and
+ * ghost. The implementation of the s3-fifo algorithm provides a concrete
+ * container for the small and main queues, and the user implements the ghost
+ * queue and any required indexing.
+ *
+ * Usage
+ * =====
+ *
+ * Consider the following entry type and an entry container registry:
+ *
+ *     struct entry {
+ *         std::string address;
+ *         std::string description;
+ *         const char* image;
+ *     };
+ *
+ *     std::map<std::string, std::unique_ptr<entry>> entries;
+ *
+ * The first step to transforming this example to provide a bounded size cache
+ * using the s3-fifo eviction algorithm is to add a control structure to the
+ * entry that stores internal metadata and provides communication between the
+ * application and the cache.
+ *
+ *     struct entry {
+ *         std::string address;
+ *         std::string description;
+ *         const char* image;
+ *
+ *         io::cache_hook hook;
+ *     };
+ *
+ * Next we'll define an instance of the cache to hold objects of type entry.
+ *
+ *     struct evict {
+ *         bool operator()(entry& e) noexcept {
+ *             e.address.clear();
+ *             e.description.clear();
+ *             free(e.image);
+ *             return true;
+ *         }
+ *     };
+ *
+ *     using cache_type = io::cache<entry, &entry::hook,
+ *       evict, io::default_cache_cost>;
+ *
+ * The `evict` callable implements the application-specific eviction operation,
+ * which in this case frees allocated memory. The type `io::default_cache_cost`
+ * provides a default cost function useful to specifiying cache capacity in
+ * terms of the number of entries in the cache.
+ *
+ * Finally, instantiate the cache with a maximum capacity and the desired size
+ * of the small queue (10% is a reasonable choice).
+ *
+ *     io::cache::config config{.capacity = 100, .small_size = 10};
+ *     cache_type cache{config};
+ *
+ * Call `cache::insert` to add an entry into the cache:
+ *
+ *     auto res = entries.emplace("e0", std::make_unique<entry>(...));
+ *     cache.insert(*res.first.second);
+ *
+ * When the application interacts with an entry it should communicate this to
+ * the cache by touching the entry:
+ *
+ *     auto e = lookup("e0");
+ *     compress(e.image);
```

Comment:
Also add: "The application also must communicate that the entry is "in-use" and is non-evict-able if it wants to use it in an async context or hold a reference to it."?

Reply:
That's not entirely true. The cache doesn't hold owning references, and it's fine if it evicts something as long as the lifetime is maintained externally.

It is true that we'll have to revisit this topic when we integrate this into the page cache, but in its current form its no different than any other synchronous data structure.